### PR TITLE
Add sampling CPU profiler to Rbkit

### PR DESCRIPTION
**Pull request by emilsoman**
* Profiles with wall and cpu clocks
* Adds new event type to the protocol called  `cpu_sample`
* Bump protocol version to v2.1
* Configurable REQ and PUB ports
* Always bind to 127.0.0.1

### CODE REVIEW

File: docs/protocol.md [hunk 1]
```diff
@@ -61,7 +63,11 @@ enums used.
   'size'                   =>  8,
   'message_counter'        =>  9,
   'correlation_id'         => 10,
-  'complete_message_count' => 11
+  'complete_message_count' => 11,
```
Comment (iffyuva) [inline docs/protocol.md]: i feel separating enums based on type of event is good idea, rather than having 1 enum.

File: experiments/rbkit_command_test.rb [hunk 2]
```diff
@@ -10,7 +10,11 @@
   'stop_memory_profile',
   'objectspace_snapshot',
   'trigger_gc',
-  'handshake'
+  'handshake',
+  'use_cpu_time',
```
Comment (iffyuva) [inline experiments/rbkit_command_test.rb]: is there a way to get current cpu profiling type?

File: ext/rbkit_event.c [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #include "rbkit_event.h"
 
-static size_t correlation_id = 0;
+static int correlation_id = 0;
```
Comment (iffyuva) [inline ext/rbkit_event.c]: why did you change this to `int` ?

File: ext/rbkit_event.c [hunk 4]
```diff
@@ -65,6 +66,14 @@ rbkit_object_space_dump_event *new_rbkit_object_space_dump_event(rbkit_object_du
   return event;
 }
 
+rbkit_cpu_sample_event *new_rbkit_cpu_sample_event(rbkit_cpu_sample *sample) {
+  rbkit_cpu_sample_event *event = malloc(sizeof(rbkit_cpu_sample_event));
```
Comment (iffyuva) [inline ext/rbkit_event.c]: can we do an assert here after malloc?

File: experiments/rbkit_command_test.rb (hunk 2 above)
Comment (emilsoman) [inline experiments/rbkit_command_test.rb]: Yeah. You'll get this along with the handshake.

File: ext/rbkit_event_packer.c [hunk 5]
```diff
@@ -207,6 +211,52 @@ static void pack_object_space_dump_event(rbkit_object_space_dump_event *event, m
...
+
+  // Keys 1 & 2 - event type and timestamp
+  pack_event_header(packer, event->event_header.event_type);
+
+  // Key 3 : Payload
+  msgpack_pack_int(packer, rbkit_message_field_payload);
+  // Value 3: Array of samples
+  msgpack_pack_array(packer, sample->frame_count);
+
+  if(sample->frame_count != 0) {
```
Comment (iffyuva) [inline ext/rbkit_event_packer.c]: no need of if statement. for loop written properly should take care of this.
`for (size_t count = 0; count != sample->frame_count; ++count)`

File: ext/rbkit_event.c (hunk 4 above)
Comment (emilsoman) [inline ext/rbkit_event.c]: We moved away from asserts previously because we don't want to kill the process if rbkit faces exceptions. Maybe we can handle such cases more gracefully, but i'm not sure what's a graceful way to handle these kind of cases.

File: CHANGELOG.md [hunk 6]
```diff
@@ -21,3 +21,7 @@ WIP
 * Send server and protocol versions along with handshake response [#103](https://github.com/code-mancers/rbkit/pull/103)
 * Split object dump into smaller messages [#77](https://github.com/code-mancers/rbkit/pull/77)
 * Fix bug with handshake broken because of Rbkit::VERSION not being available.
+* Make REQ and PUB ports configurable using ENV variables
+* Always bind to 127.0.0.1
+* Add CPU Sampling profiler and cpu_sample message type. Bump up protocol version to v2.1
+* Use C89 std for compiling rbkit
```
Comment (iffyuva) [inline CHANGELOG.md]: why did we move back to c89?

File: docs/protocol.md (hunk 1 above)
Comment (emilsoman) [inline docs/protocol.md]: We have many enums repeated in multiple events. For example file, line number, payload, timestamp, event_type appear in other event types also.

File: CHANGELOG.md (hunk 6 above)
Comment (emilsoman) [inline CHANGELOG.md]: In our dev boxes C99 is what gets used for compilation, but in travis, it uses C89 and gives compilation warnings. I thought we should stick to one std , so I did some research to see which one to use. Looks like Ruby-core officially supports only C89 std, also C89 is the most widely implemented standard.

File: ext/rbkit_event.c (hunk 3 above)
Comment (emilsoman) [inline ext/rbkit_event.c]: Max value of `correlation_id` = the total number of object space dumps i made after running the server. I thought int would be better. Well, we can keep size_t if you think it's better, not a biggie.
